ZIP 317 changes (#640)
zcash/zips: zip-0317.rst

Let :math:`\mathsf{ceiling}(x)` be the smallest integer :math:`\geq x`.

Fee calculation
---------------

This specification defines several parameters that are used to calculate the
conventional fee:

===================================== ==============================================
Parameter                             Units
===================================== ==============================================
:math:`marginal\_fee`                 zatoshis per logical action (as defined below)
:math:`grace\_actions`                logical actions
:math:`p2pkh\_standard\_input\_size`  bytes
:math:`p2pkh\_standard\_output\_size` bytes
===================================== ==============================================

Wallets implementing this specification SHOULD use a conventional fee
calculated in zatoshis per the following formula:

.. math::

   \begin{array}{rcl}
     logical\_actions  &=& \mathsf{max}\big(\mathsf{ceiling}\big(\frac{tx\_in\_total\_size}{p2pkh\_standard\_input\_size}\big),
                                            \mathsf{ceiling}\big(\frac{tx\_out\_total\_size}{p2pkh\_standard\_output\_size}\big)\big) \;+ \\

>                        & & 2 \cdot nJoinSplits \;+ \\

Author: ```suggestion
                       & & 2 \cdot nJoinSplit \;+ \\
```

                       & & \mathsf{max}(nSpendsSapling, nOutputsSapling) \;+ \\
                       & & nActionsOrchard \\
     conventional\_fee &=& marginal\_fee \cdot \mathsf{max}(grace\_actions, logical\_actions)
   \end{array}

The inputs to this formula are taken from transaction fields defined in the Zcash protocol

> specification [#protocol-txnencoding]:

Author: ```suggestion
specification [#protocol-txnencoding]_:
```

============================ ====== ===========================================
Input                        Units  Description
============================ ====== ===========================================
:math:`tx\_in\_total\_size`  bytes  total size in bytes of the ``tx_in`` field
:math:`tx\_out\_total\_size` bytes  total size in bytes of the ``tx_out`` field
:math:`nJoinSplit`           number the number of Sprout JoinSplits
:math:`nSpendsSapling`       number the number of Sapling spends
:math:`nOutputsSapling`      number the number of Sapling outputs
:math:`nActionsOrchard`      number the number of Orchard actions
============================ ====== ===========================================

The parameters are set to the following values:

* :math:`marginal\_fee = 5000`;
* :math:`grace\_actions = 2`;

> * :math:`p2pkh\_standard\_input\_size = TODO` bytes;
> * :math:`p2pkh\_standard\_output\_size = TODO` bytes.

Reviewer: ```suggestion
* :math:`p2pkh\_standard\_input\_size = 150` bytes;
* :math:`p2pkh\_standard\_output\_size = 34` bytes.
```

Reviewer 2: For per-input sizes the values can range from 147 to 149 bytes due to ❤️‍🔥 just ❤️‍🔥 how ❤️‍🔥 lovely ❤️‍🔥 the DER encoding of a secp256k1 signature is. 150 bytes seems fine for this purpose.

significant per-logical-action cost advantage by using transactions
with a smaller or larger number of logical actions.

Transparent Contribution
~~~~~~~~~~~~~~~~~~~~~~~~

The specified formula calculates the contribution of transparent inputs
and outputs based on their total size relative to a typical input or
output. Another considered approach was to calculate this contribution
simply as :math:`\mathsf{max}(transparent\_inputs, transparent\_outputs)`.
However, this would allow a denial-of-service adversary to create
transactions with transparent components containing arbitrarily large
scripts.

The chosen values for :math:`p2pkh\_standard\_input\_size` and
:math:`p2pkh\_standard\_output\_size` are based on the maximum encoded

> length for simple P2PKH inputs and outputs, as follows:

Author: ```suggestion
length for P2PKH inputs and outputs, as follows:
```

* :math:`p2pkh\_standard\_input\_size`

  * outpoint: 36 bytes

>   * stack script: 72 bytes

Author: ```suggestion
  * stack script (including length): 72 bytes
```

>   * stack script: 72 bytes
>   * redeem script: 34 bytes

Author: ```suggestion
  * stack script (including length): 72 bytes
  * redeem script (including length): 34 bytes
```

  * sequence: 4 bytes

* :math:`p2pkh\_standard\_output\_size`

  * value: 8 bytes

>   * script: 26 bytes

Author: ```suggestion
  * script (including length): 26 bytes
```

> Simple P2SH inputs and outputs are not typically larger than this.

Author: ```suggestion
Simple P2SH outputs are not typically larger than this, but P2SH inputs may be larger.
It is intended that such P2SH inputs count as more than one logical action.
```

change to the :math:`low\_fee\_penalty` threshold described in
`Mempool size limiting`_ immediately.

Miners can deploy restrictions to their policies for transaction inclusion,
once a sufficient proportion of wallets in the ecosystem are observed to

> be paying at least the updated conventional transaction fee.

Reviewer 3: Should this say "proportion of wallets" or "proportion of transactions"? The latter is more easily observable by looking at the blockchain.

Reviewer 4: I'm +1 on "proportion of transactions".
